### ISSUE TITLE
Can I use pnpjs version 3.X for SharePoint 2013 on prem?

### ISSUE THREAD

**Issue by AkumaTen** [pnp/pnpjs#2205]
Dear experts,
we developed a framework using pnpjs to add client side web parts to SharePoint 2013. We are facing a transition to the cloud and were able to re-use what have done so far in conjunction with SPFx. Basically our base code is serving both SP 2013 on prem (production) and SPO (POC environment at this stage).

I've read the pnpjs documentation but when you say "If you are running in 2016 or 2019 on-premises you will need to use version 2", it is not clear to me if this is mandatory or advised. If it is mandatory, till when do you think will SPO work with pnpjs V2.x?

To put it in other words, if we upgrade our code to work with version 3, will  this still work on 2013 on prem?

Our target infrastructure will be and hybrid cloud SP 2016 + SPO but at the moment SP 2013 is still important for us.

### Category
- [ ] Enhancement
- [ ] Bug
- [x] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: 2.13.0

Please specify what version(s) of SharePoint you are targeting: 2013/online

Thank you for your feedback!

**Comment by patrick-rodgers** [comment]
You can try - but that isn't a scenario we've tested or are looking to support. The library, through observers, should allow you to customize the request pipeline. One issue you might encounter is that we've switched to minimal metadata that not all sp2013 environments support - meaning our request bodies are no longer expressed in verbose.

**Comment by AkumaTen** [comment]
migrating to SP 2016 will help? You said "not all sp203 environment support", so what have I to check? On prem we can change whatever we want.

Can you also tell me more about he pipeline please? How would this help me in targeting SP 2013?

let's say we keep V2.x for SPO, is this something that can work in your opinion or if it works today, will it break soon?

**Comment by juliemturner** [comment]
I can tell you, because I tested it, that you cannot really use V3 with a version of SPFx before v1.12.1... although I didn't work very hard to "make" it work the upshot is that v3 uses TypeScript 4 which isn't natively supported by SPFx... back until 1.12.1 you can fix this by updating the version of the rush-stack-compiler but before that I'm not sure if you could make it work.  Given that to target SP2016 on-prem you cannot use a version of SPFx newer than 1.1 (I think, didn't verify my memory) this is a non-starter.  

So, long story short, in my opinion, your best way forward is to build libraries of functionality instead, host those bundled libraries in the cloud (like an Azure CDN) and then bootstrap them into SPFx or classic script editor web parts for your on-prem environment and then bootstrap them into SPFx 1.14 of Online... Now you have 1 codebase and you can use modern tooling.

**Comment by AkumaTen** [comment]
I get your point but our scenario is different. We do use SPFx v1.12.1. e we do use the rush-stack-compiler to enable a certain version of TS. We will not use SPFx within SP 2016 because of the architecture of our framework.  Please think of SP 2016 as a simple replacement of SP 2013 for legacy stuff. 

So, what do you suggest? should I take the effort to adopt V3 in the hopes of having my rest calls to SP2013/SP2016 continuing to work, or  better should I keep the V2 because it is still fine with SPO?

**Comment by juliemturner** [comment]
Direct REST calls have not changed on the platform (including SPO) other than the point that @patrick-rodgers made about minimal metadata in 2013. You can use V2 if you want, there's no reason you need to move to V3, based on your question the only difference between V3 is the coding environment not the calls themselves other than the point about minimal metadata.

**Comment by AkumaTen** [comment]
Thanks! I think we'll go for keeping the V2. I don't know what minimal metadata is, so, can you please suggest me what to do on SP 2013 in case we would try the V3?

Regarding V2: that means you will keep on updating it following new features that could come into SPO, am I correct? do yo have an idea when will the migration to V3 be mandatory?  this will help us to better choose where and when to put our effort

**Comment by AkumaTen** [comment]
for us, maintaining V2 would be safe in short term but would be risky in long term because a change in SPO could break V2. Is this correct? 

Is it possible to change the way V3 manage the Accept header in case we want to try V3 against SP 2013 and  find problems?

**Comment by AkumaTen** [comment]
thanks all for the support :)
